### DISCUSSION TITLE
[HELP] - Use of indexof

### DISCUSSION THREAD

**Discussion by hiranokite** [moudey/Shell#234] [Q&A]
I'm trying to set positions to some entries using `indexof`, mentioned here #229, but I'm facing some difficultly.

`item(type='file|dir|back.dir|drive' mode="multiple" title='TEST' image=[\uE280,#f00])`

this menu will appear on many occasions, but I'm not manage it to work like i want.

Need to appears after `Open with`, while file is selected, `position=indexof("Open with", 1)`
but on files like `.exe` it wont appears since executable only do option to `Open`, so will need a `fallback` in case `Open with` isn't available it try on another menu, maybe possible using if(), but since i have not found the documentation of `indexof` it's became difficulty, every time i try a new code it not work as intend or Shell just fail to load because the syntax error.

I need it to work in the following order, Open with > Open > Refresh

Open with = file
Open = file, folder and drive
Refresh = folder background

any help or suggestion will be welcome.

File
![image](https://user-images.githubusercontent.com/5361459/231676890-d5589ae1-dbba-4a1f-b984-20026f7c3862.png)

EXE
![image](https://user-images.githubusercontent.com/5361459/231676751-f2068ece-add3-47e5-ab1c-ac2885ba5114.png)

Directory
![image](https://user-images.githubusercontent.com/5361459/231675981-1ec0b6e5-770b-447e-a771-be21a1ef41e9.png)

Dir Background
![image](https://user-images.githubusercontent.com/5361459/231675130-4f7c7ce4-d7f6-45ad-8254-ee157d73efdc.png)

Drive
![image](https://user-images.githubusercontent.com/5361459/231675561-6fded627-afc2-4a94-b9c7-e394d6513537.png)

**Comment by moudey** [comment]
```
indexof('item name' [ [, index] [, fallback index] ])

indexof('item name')
indexof('item name', 1)
indexof('item name', 1, 2)
```

This is a temporary solution
```
item(type='file' find='!.exe' mode="multiple" title='TEST' image=[\uE280,#f00] position=indexof("Open with", 1))
item(type='file' find='.exe' mode="multiple" title='TEST' image=[\uE280,#f00] position=indexof("Open", 1))
item(type='dir|drive' mode="multiple" title='TEST' image=[\uE280,#f00] position=indexof("Open", 1))
item(type='back.dir' title='TEST' image=[\uE280,#f00] position=indexof("Refresh", 1))
```

I will support this function with multiple strings like the following example
```
indexof('open with|"open"|"edit"'], 1, 2)
```

**Reply by hiranokite** [reply]
is possible to use id functions with indexof?
like id.open_with.title or partial string "Open wi*"

**Reply by moudey** [reply]
Try this [build](https://nilesoft.org/download/shell/debug.zip)

<i>Note:</i> That there are changes in the structure of the configuration file, where the variables, static and dynamic section has been dropped and the theme section has been moved to the top level statement.

The configuration file is made simpler this way:
```
// Global variable
$variable='test var1'
$variable2='test var2'

// image declaration
@image1=[\e080, #f00]

settings 
{
    $variable4='test var4'
    ....
}

theme
{
    $variable5='test var5'
    ...
}

// modified items
modify(...)

// new items
menu()
{
    // Private variable
    $variable3='test var3'
    ...
}

item(...)
```

**Reply by moudey** [reply]
> is possible to use id functions with indexof? like id.open_with.title or partial string "Open wi*"

Yes they can be used
```
indexof('@id.open_with.title|"@id.open.title"', 1, "middle")
```

**Reply by hiranokite** [reply]
I'll try, but seem shell on taskbar is broken again on Windows 11 on this build.

**Reply by moudey** [reply]
> I'll try, but seem shell on taskbar is broken again on Windows 11 on this build.

Don't worry about that because I'm trying to improve the handling of the taskbar and reduce the use of hook functions as much as possible. Try it now, you will find it works

**Reply by hiranokite** [reply]
worked perfectly with string, but not worked with id, or I'm doing some mess?
`position=indexof('"Open with"|"@id.open.title"|"Open in new window"|"Refresh"', 1)`
using this code `Open` is ignored and the custom menu entry show at the bottom.

**Reply by moudey** [reply]
`id.open.title` = `&Open`
This issue has been resolved by adding this function `id.open.name` with build 16
`id.open.name` = `Open`

**Reply by hiranokite** [reply]
id working on build 16, but for some reason `id.open.name` is overwriting others ids
`position=indexof('"@id.open_in_new_window.name"|"@id.open_with.name"|"@id.refresh.name"|"@id.open.name"', 1)`

**Reply by moudey** [reply]
```
position=indexof('
    "@id.open_in_new_window.name"|"@id.open_with.name"|"@id.refresh.name"
    @if(sel.file.ext==".exe",'|"@id.open.name"')', 1)
```

**Reply by hiranokite** [reply]
> ```
> position=indexof('
>     "@id.open_in_new_window.name"|"@id.open_with.name"|"@id.refresh.name"
>     @if(sel.file.ext==".exe",'|"@id.open.name"')', 1)
> ```

with this solution, the item goes to the bottom on folder, folder-shortcut, exe-shortcuts and url-shortcuts

**Reply by moudey** [reply]
The shortcut is treated as a file

folder-shortcut and file-shortcuts == '.lnk'
```
position=indexof('
    "@id.open_in_new_window.name"|"@id.open_with.name"|"@id.refresh.name"
    @if(sel.file.ext==".exe" || sel.file.ext==".lnk",'|"@id.open.name"')', 1)
```

**Reply by hiranokite** [reply]
ok, following you syntax I did this, to add folder and url-shortcut, seem now all is working
`position=indexof('"@id.open_in_new_window.name"|"@id.open_with.name"|"@id.refresh.name"@if(sel.file.ext==".exe"||sel.file.ext==".lnk"||sel.file.ext==".url"||sel.directory,'|"@id.open.name"')', 1)`

**Reply by moudey** [reply]
The indexof function has been improved as it gives priority from left to right. build 17

**Reply by hiranokite** [reply]
> The indexof function has been improved as it gives priority from left to right. build 17

this will help a ton, thank you!

**Reply by hiranokite** [reply]
@moudey I'm experiencing strange behavior when using `indexof` between items in the windows itself and items created from the shell.

some examples

Default
![image](https://user-images.githubusercontent.com/5361459/233536873-3a9b042f-c883-42d1-8681-85bea809e6a5.png)

Shell created items work fine
`item(title="item1" image=#FF0000 position=indexof("view", 1))`
`item(title="item2" image=#FAF000 position=indexof("item1", 0))`
![image](https://user-images.githubusercontent.com/5361459/233537771-d270c9bb-793a-4a5f-af92-3645f23da650.png)

now trying to modify using windows default items
`modify(find="refresh" position=indexof("view", 0))`
Refresh work with View
`modify(find="new" position=indexof("item1", 1))`
But New seems can't find item1, and it goes all the way down
![image](https://user-images.githubusercontent.com/5361459/233538552-86102166-facf-4c7f-b3ad-19ed87905bc8.png)

**Reply by moudey** [reply]
`indexof` has been improved, [build 24](https://nilesoft.org/download/shell/debug.zip)

**Reply by hiranokite** [reply]
@moudey I'm trying to remove a menu but I'm facing this scenario.

the menu in specifically is `Sort by` on `desktop`
![image](https://user-images.githubusercontent.com/5361459/235012869-01d0ada3-cdcf-4a4f-81cf-77b6ada4c742.png)

When I uncheck `View > Show desktop icons` all items inside Sort by become grayed.
![image](https://user-images.githubusercontent.com/5361459/235012499-ebca84dc-4f49-47f6-999b-d67c28cf33cd.png)

`modify(type='*' where=this.disabled visibility=remove)`
when using this line all items inside `Sort by` become disabled, but the menu still visible.

![image](https://user-images.githubusercontent.com/5361459/235011892-00c4785d-1ac1-49ef-b4fd-7891dca98447.png)
I need to completely remove Sort by menu, but only when `Show desktop icons` is unchecked.

**Reply by moudey** [reply]
The correct expression for the value of the visibility property
```
visibility="remove"
visibility=vis.remove
visibility=0
```

**Reply by hiranokite** [reply]
> The correct expression for the value of the visibility property
> ```
> visibility="remove"
> visibility=vis.remove
> visibility=0
> ```

but there's a way to remove Sort by menu only when it's empty?

**Reply by hiranokite** [reply]
any suggestion?

**Reply by moudey** [reply]
> any suggestion?

Sorry for the delay, it will be worked on

**Comment by moudey** [comment]
The taskbar works again. It was due to an extra byte being written to the registry while the shell was enabled.

**Reply by hiranokite** [reply]
There's a bug that freezes the entirely taskbar, it happens on old version too.
I didn't find any issue reported about this.

To trigger the bug:
- Just hold any keyboard key (with `ALT` won't work) and start right clicking on Taskbar, if you do it faster you can trigger the bug with +/- 5 clicks, after that taskbar become unresponsive, until a explorer restart.

**Reply by moudey** [reply]
>Just hold any keyboard key (with ALT won't work) and start right clicking on Taskbar, if you do it faster you can trigger the bug with +/- 5 clicks, after that taskbar become unresponsive, until a explorer restart.

The issue has been fixed

**Reply by hiranokite** [reply]
~~tested on build 20, the bug still persists, but a little bit difficult to trigger.
and now navigation keys work on taskbar.~~

**Reply by hiranokite** [reply]
> > Just hold any keyboard key (with ALT won't work) and start right clicking on Taskbar, if you do it faster you can trigger the bug with +/- 5 clicks, after that taskbar become unresponsive, until a explorer restart.
> 
> The issue has been fixed

@moudey Sorry, looks like it was something to do with remnants of old ExplorerPatcher installations, test in a new VM and everything is working fine.

**Reply by hiranokite** [reply]
OK, since I was using build 14 on my main configuration, now that I ported to the new structure I made some tests and got some old bugs.

Display BUG returned at build 15-20.
https://github.com/moudey/Shell/discussions/181#discussioncomment-5347478

And on build 20 i can't use the `LMB` + `RMB` to show the hide menu on taskbar anymore.
`item(visibility=key.lbutton() or key.lshift() or key.rshift() title='Restart Explorer' image=\uE1AB cmd=command.restart_explorer)`

**Reply by hiranokite** [reply]
Could you add a function to return build version value, or is that already present?
`@app.ver			//	1.8.1`
`@app.version			//	1.8.1.21`
`@app.version.build	//	21`

**Reply by moudey** [reply]
[Build 22](https://nilesoft.org/download/shell/debug.zip)

Fix
>And on build 20 i can't use the LMB + RMB to show the hide menu on taskbar anymore.

New
```
app.ver.major
app.ver.minor
app.ver.build
```

**Reply by hiranokite** [reply]
Thanks, everything is working great.

and there's something I noted since build 20, this line is appearing on the log every time I open Taskbar context menu, does that mean something?

`2023-04-18 02:24:38 [info] ************`

**Reply by moudey** [reply]
It was just a test when I clicked on the taskbar I forgot to remove it.

**Reply by hiranokite** [reply]
Understood, thought it was some error on my windows.

**Reply by hiranokite** [reply]
> i can't use the `LMB` + `RMB` to show the hide menu on taskbar anymore. `item(visibility=key.lbutton() or key.lshift() or key.rshift() title='Restart Explorer' image=\uE1AB cmd=command.restart_explorer)`

hello @moudey, this behavior come back at new build 33.
since this shortcut it's simple to use I use it a lot. can you take a look?

**Comment by moudey** [comment]
This issue is mainly in the taskbar. Try the default context menu you will find the same problem. I will look for a solution to it.
